### PR TITLE
Player Indicators: plugin rework

### PR DESCRIPTION
**Pull request by pklite**
current todos:
- [x] re-add caller highlighting
- [x] re-add old caller target highlighting
- [x] re-add uncharged glory 
- [ ] re-add agility shortcut stuff
- [x] few other things i can't remember atm but will update later
- [x] etc, etc

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java [hunk 1]
```diff
@@ -1,298 +1,124 @@
...
- * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
- * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
- * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
- * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
- * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
- * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
- * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
- * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
- */
+/*******************************************************************************
```
Comment (pklite) [inline runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java]: wtf why did intellij do this lmao
another todo: fix copyright

**Comment by chrisjohnson** [comment]
How did this get merged with "todo:  re-add agility shortcut stuff" still left undone? Can you add that feature back in please?

**Comment by Ganom** [comment]
We wanted to merge it before the update that was pushed later that day. I will be added back in, @pklite is still making significant changes.

**Comment by chrisjohnson** [comment]
How can people who don't know how to build from scratch get a previous version of RL+ before this feature was ripped out? I've got a clan event tonight and nobody else knows how to build so I gotta get them all older versions.

What was the motivation for rushing this change through before it was complete?

**Comment by chrisjohnson** [comment]
Seems like this PR also broke friend/clan indicators taking priority over other/attackable.

**Comment by devOSRS** [comment]
https://github.com/runelite-extended/runelite/issues/1485

**Comment by pklite** [comment]
@chrisjohnson 
The agility features called a high score lookup for every local player on every single frame without even caching the results, with 10 players on screen at 50 fps at the very least that's 500 request attempts per second. Decided to remove it based on that until we  can think of an alternative way to do it, possibly location based. Or maybe, do you only need to know if they can use the shortcuts if they are tbed and also within the rev caves?

**Comment by chrisjohnson** [comment]
No it doesn't, take a look at the class it's calling: https://github.com/runelite-extended/runelite/blob/master/runelite-client/src/main/java/net/runelite/client/game/HiscoreManager.java#L60-L66

It caches the lookup, I confirmed this. I also addressed it in the PR comments when @ThatGamerBlue asked about the same thing: https://github.com/runelite-extended/runelite/pull/1100#discussion_r307172584

Rather than just silently remove it with no notification to myself, why not simply reach out to me about it if you needed clarification?

**Comment by Ganom** [comment]
> No it doesn't, take a look at the class it's calling: https://github.com/runelite-extended/runelite/blob/master/runelite-client/src/main/java/net/runelite/client/game/HiscoreManager.java#L60-L66
> 
> It caches the lookup, I confirmed this. I also addressed it in the PR comments when @ThatGamerBlue asked about the same thing: [#1100 (comment)](https://github.com/runelite-extended/runelite/pull/1100#discussion_r307172584)
> 
> Rather than just silently remove it with no notification to myself, why not simply reach out to me about it if you needed clarification?

It will be added back, we already said that. We've just been very busy. I will do it today.

**Comment by chrisjohnson** [comment]
I'm not whining about getting it back in, I know you said that which is why I didn't bring it back up again. I was just responding to the suggestion that the lookup wasn't cached, and suggesting communication as as a first step moving forward.

Please don't take offense, I'm not trying to be rude, just understand what happened

**Comment by pklite** [comment]
Yeah my bad, I didn't even check the Highscore manage cache. Just reverted the changes and looked into it now.I tested to see how effective the cache is when its being hit with so many checks since it has such a small size by dumping its stats while hopping a few worlds in edgeville and bypassing the wildy check. it's not very effective at all in this case, and on every world hop it becomes dramatically less efficient

**Comment by pklite** [comment]
`2019-08-31 12:59:27 [Client] INFO  n.r.c.p.p.PlayerIndicatorsOverlay - [Runtime 00:00:45]: hitRate: 0.0 average load penalty: 6.885638106229267E-5ms cache removals: 696535, total load time: 47ms 
2019-08-31 12:59:27 [Client] INFO  n.r.c.p.p.PlayerIndicatorsOverlay - [Runtime 00:00:45]: hitRate: 0.0 average load penalty: 6.883502615880947E-5ms cache removals: 697916, total load time: 48ms 
2019-08-31 12:59:27 [Client] INFO  n.r.c.p.p.PlayerIndicatorsOverlay - [Runtime 00:00:45]: hitRate: 0.0 average load penalty: 6.881542244108839E-5ms cache removals: 699437, total load time: 48ms 
2019-08-31 12:59:32 [Client] INFO  n.r.c.p.p.PlayerIndicatorsOverlay - [Runtime 00:00:50]: hitRate: 1.300105865763355E-4 average load penalty: 6.940520281430931E-5ms cache removals: 699724, total load time: 48ms 
2019-08-31 12:59:32 [Client] INFO  n.r.c.p.p.PlayerIndicatorsOverlay - [Runtime 00:00:51]: hitRate: 4.924406787402654E-4 average load penalty: 6.961839893637532E-5ms cache removals: 700119, total load time: 48ms 
2019-08-31 12:59:33 [Client] INFO  n.r.c.p.p.PlayerIndicatorsOverlay - [Runtime 00:00:51]: hitRate: 7.773520962832585E-4 average load penalty: 6.968133888513789E-5ms cache removals: 700425, total load time: 48ms 
2019-08-31 12:59:33 [Client] INFO  n.r.c.p.p.PlayerIndicatorsOverlay - [Runtime 00:00:51]: hitRate: 0.0010405337795750347 average load penalty: 6.972825337847961E-5ms cache removals: 700705, total load time: 48ms 
2019-08-31 12:59:33 [Client] INFO  n.r.c.p.p.PlayerIndicatorsOverlay - [Runtime 00:00:51]: hitRate: 0.0011314957690038562 average load penalty: 6.972782238505305E-5ms cache removals: 700804, total load time: 48ms 
2019-08-31 12:59:33 [Client] INFO  n.r.c.p.p.PlayerIndicatorsOverlay - [Runtime 00:00:51]: hitRate: 0.0012167270043582762 average load penalty: 6.973101398087668E-5ms cache removals: 700901, total load time: 48ms 
2019-08-31 12:59:33 [Client] INFO  n.r.c.p.p.PlayerIndicatorsOverlay - [Runtime 00:00:51]: hitRate: 0.0023455336124485185 average load penalty: 7.00257233994076E-5ms cache removals: 702112, total load time: 49ms
2019-08-31 12:59:33 [Client] INFO  n.r.c.p.p.PlayerIndicatorsOverlay - [Runtime 00:00:52]: hitRate: 0.003807679622685844 average load penalty: 0.0016814382390469723ms cache removals: 703649, total load time: 1183ms 
2019-08-31 12:59:34 [Client] INFO  n.r.c.p.p.PlayerIndicatorsOverlay - [Runtime 00:00:52]: hitRate: 0.004406905187936556 average load penalty: 0.001680001594244253ms cache removals: 704280, total load time: 1183ms 
2019-08-31 12:59:34 [Client] INFO  n.r.c.p.p.PlayerIndicatorsOverlay - [Runtime 00:00:52]: hitRate: 0.0048509833265095795 average load penalty: 0.0016789444270260685ms cache removals: 704746, total load time: 1183ms 
2019-08-31 12:59:34 [Client] INFO  n.r.c.p.p.PlayerIndicatorsOverlay - [Runtime 00:00:52]: hitRate: 0.005459643325857395 average load penalty: 0.0016775279887741579ms cache removals: 705386, total load time: 1183ms 
2019-08-31 12:59:34 [Client] INFO  n.r.c.p.p.PlayerIndicatorsOverlay - [Runtime 00:00:52]: hitRate: 0.005934713922852943 average load penalty: 0.0016763940332627965ms cache removals: 705885, total load time: 1183ms `

**Comment by pklite** [comment]
`2019-08-31 13:01:43 [Client] INFO  n.r.c.p.p.PlayerIndicatorsOverlay - 
[Runtime 00:05:01]:
 hitRate: 0.2958699121927128
 average load penalty: 3.111210609444799ms
 cache removals: 778621
 total load time: 2423007ms `
this was after 5 minutes and 2 or 3 world hops. i think if we creating a custom caching method that isn't done on every frame similar to the player scouter would be a good way to do this. Could also try increasing the cache size which would help with load time as runtime gets longer but it would also use more ram. might still be worth looking into though

**Comment by ThatGamerBlue** [comment]
Just give it a `Map<String, Integer> agilLevels = new SomeFormOfEfficientMapImplementation<>();`

**Comment by chrisjohnson** [comment]
Rather than re-invent the wheel what do you think about modifying
HiscoreManagers cache to behave more like what you're talking about. Then
we get free lunch anywhere we want to look up high scores regularly. We
could even update anything else that needs high scores to use
HiscoreManager if they aren't already to get a global shared cache instead
of plugin localized one.

if the cash in that class is not behaving for us then it seems like that is
the culprit to address rather than cargo culting a hacky Band-Aid from
plug-in to plug in. Thoughts?

**Comment by chrisjohnson** [comment]
In my testing every subsequent call for the same player amounted to a cache
hit, which if the cache is tuned right certainly ought to be good enough to
call every frame

**Pull request merged**